### DISCUSSION TITLE
Updating lists dynamically

### DISCUSSION THREAD

**Discussion by KimSilvaA** [bonsai-rx/bonsai#1322] [Q&A]
Hello,

I'm using Bonsai for a behavior task I'm running and have most of it completed. I have it set up so that when a trial ends, I know whether the mouse got it correct or not using a Python Transform. Now I'd like to create a list that stores every value but I don't know how to create and update a list each time a trial ends. Ultimately, I want to store the values and then feed it into a python transform that will tell me when to stop the experiment. Any advice is greatly appreciated!

![Capture](https://user-images.githubusercontent.com/124469197/231847989-4c3c55c3-1791-4e54-bde4-e0968137a57a.PNG)

**Comment by bruno-f-cruz** [comment]
Can you share the workflow file? thanks

**Reply by KimSilvaA** [reply]
Sure. 
![image](https://user-images.githubusercontent.com/124469197/231865374-85d87655-4144-4a98-b500-91a30ca2eb5c.svg)

**Reply by bruno-f-cruz** [reply]
Can you please share the `.bonsai` workflow file as instructed in #864 Thanks!

**Reply by KimSilvaA** [reply]
[Bonsai.zip](https://github.com/bonsai-rx/bonsai/files/11225970/Bonsai.zip)

Here it is! Thanks

**Reply by bruno-f-cruz** [reply]
I tend to prefer using replay subjects for this porpose, but if you really must use list or arrays here s a possible solution (@glopesdev I wonder if there is a bonsai operator that implements the List.Add() method that I am missing perhaps?)

![image](https://user-images.githubusercontent.com/7049351/231972234-94fe5cef-043a-46be-bafe-997e84c01ee7.png)

Inside the `PreprendToList(SelectMany)`
![image](https://user-images.githubusercontent.com/7049351/231972293-9bde82de-585b-4049-afae-581ddfff46fc.png)

<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml

<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.7.3"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns:wie="clr-namespace:Bonsai.Windows.Input;assembly=Bonsai.Windows.Input"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="IntProperty">
          <Value>0</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Take">
          <rx:Count>1</rx:Count>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:ToList" />
      </Expression>
      <Expression xsi:type="rx:BehaviorSubject">
        <Name>MyCurrentList</Name>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="wie:KeyDown">
          <wie:Filter>None</wie:Filter>
          <wie:SuppressRepetitions>false</wie:SuppressRepetitions>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="IntProperty">
          <Value>1</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="rx:SelectMany">
        <Name>PreprendToList</Name>
        <Workflow>
          <Nodes>
            <Expression xsi:type="SubscribeSubject">
              <Name>MyCurrentList</Name>
            </Expression>
            <Expression xsi:type="Combinator">
              <Combinator xsi:type="rx:Take">
                <rx:Count>1</rx:Count>
              </Combinator>
            </Expression>
            <Expression xsi:type="Combinator">
              <Combinator xsi:type="rx:Merge" />
            </Expression>
            <Expression xsi:type="WorkflowInput">
              <Name>Source1</Name>
            </Expression>
            <Expression xsi:type="Combinator">
              <Combinator xsi:type="rx:Take">
                <rx:Count>1</rx:Count>
              </Combinator>
            </Expression>
            <Expression xsi:type="Combinator">
              <Combinator xsi:type="rx:Concat" />
            </Expression>
            <Expression xsi:type="Combinator">
              <Combinator xsi:type="rx:ToList" />
            </Expression>
            <Expression xsi:type="WorkflowOutput" />
          </Nodes>
          <Edges>
            <Edge From="0" To="1" Label="Source1" />
            <Edge From="1" To="2" Label="Source1" />
            <Edge From="2" To="5" Label="Source2" />
            <Edge From="3" To="4" Label="Source1" />
            <Edge From="4" To="5" Label="Source1" />
            <Edge From="5" To="6" Label="Source1" />
            <Edge From="6" To="7" Label="Source1" />
          </Edges>
        </Workflow>
      </Expression>
      <Expression xsi:type="MulticastSubject">
        <Name>MyCurrentList</Name>
      </Expression>
      <Expression xsi:type="SubscribeSubject">
        <Name>MyCurrentList</Name>
      </Expression>
      <Expression xsi:type="MemberSelector">
        <Selector>Count</Selector>
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="1" Label="Source1" />
      <Edge From="1" To="2" Label="Source1" />
      <Edge From="2" To="3" Label="Source1" />
      <Edge From="4" To="5" Label="Source1" />
      <Edge From="5" To="6" Label="Source1" />
      <Edge From="6" To="7" Label="Source1" />
      <Edge From="8" To="9" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>

```

</details>

**Reply by KimSilvaA** [reply]
> I tend to prefer using replay subjects for this porpose, but if you really must use list or arrays here s a possible solution (@glopesdev I wonder if there is a bonsai operator that implements the List.Add() method that I am missing perhaps?)
> 
> ![image](https://user-images.githubusercontent.com/7049351/231972234-94fe5cef-043a-46be-bafe-997e84c01ee7.png)
> 
> Inside the `PreprendToList(SelectMany)` ![image](https://user-images.githubusercontent.com/7049351/231972293-9bde82de-585b-4049-afae-581ddfff46fc.png)
> 
> Workflow (Copy+Paste me)

Thank you so much for the possible solution! I'm still new to Bonsai which is why I was thinking of creating lists but I'll try to play around with ReplaySubject. I really appreciate the help :)

**Comment by KimSilvaA** [comment]
Hi again @bruno-f-cruz. I've been trying to use the workflow you posted without success so I've been trying to use ReplaySubject. Would this node allow me to index the numbers I've stored? Thanks in advance!

**Reply by KimSilvaA** [reply]
I ended up using WindowCount if anyone else is doing something similar!
